File: SeamlessPayCore/Classes/SPPaymentViewController.m
/**
 * Copyright (c) Seamless Payments, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */


#import "SPPaymentViewController.h"


@interface SPPaymentViewController () <UITextFieldDelegate>


@end

@implementation SPPaymentViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.view.backgroundColor = [UIColor whiteColor];
    
    SPPaymentCardTextField *cardTextField = [[SPPaymentCardTextField alloc] init];
    cardTextField.postalCodeEntryEnabled = TRUE;
    cardTextField.countryCode = @"US";
    self.cardTextField = cardTextField;
    
    UITextField *amountTextField = [[UITextField alloc] initWithFrame:CGRectMake(70, 0, 150, 22)];
    amountTextField.text = [self formatedValue:@([self.paymentAmount floatValue]*100) ?: @(0)];
    amountTextField.keyboardType = UIKeyboardTypeNumberPad;
    amountTextField.delegate = self;
    self.amountTextField = amountTextField;
    
    UIButton *button = [UIButton buttonWithType:UIButtonTypeCustom];
    button.layer.cornerRadius = 5;
    button.backgroundColor = [UIColor systemBlueColor];
    button.titleLabel.font = [UIFont systemFontOfSize:22];
    [button setTitle:@"Pay" forState:UIControlStateNormal];
    [button addTarget:self
               action:@selector(pay)
     forControlEvents:UIControlEventTouchUpInside];
    self.payButton = button;
    
    UILabel *amountLabel = [[UILabel alloc] initWithFrame:CGRectMake(0, 0, 250, 30)];
    amountLabel.userInteractionEnabled = _isEditAmount;
    amountLabel.textColor = [UIColor darkGrayColor];
    amountLabel.text = @"Amount:";
    [amountLabel addSubview:amountTextField];
    self.amountLabel = amountLabel;
    
    UILabel *descriptionLabel = [UILabel new];
    descriptionLabel.text = [_paymentDescription copy];
    descriptionLabel.font = [UIFont systemFontOfSize:13];
    descriptionLabel.textColor = [UIColor grayColor];
    descriptionLabel.numberOfLines = 2;
    [descriptionLabel sizeToFit];
    self.descriptionLabel = descriptionLabel;
    
    UIImageView *logoImageView = [[UIImageView alloc] initWithImage:self.logoImage];
    logoImageView.contentMode = UIViewContentModeCenter;

    UIStackView *stackView = [[UIStackView alloc]
                              initWithArrangedSubviews:@[logoImageView, amountLabel, descriptionLabel, cardTextField, button ]];
    stackView.axis = UILayoutConstraintAxisVertical;
    stackView.translatesAutoresizingMaskIntoConstraints = FALSE;
    stackView.spacing = 20;
    [self.view addSubview:stackView];
    
    [NSLayoutConstraint activateConstraints:@[
        [stackView.leftAnchor
         constraintEqualToSystemSpacingAfterAnchor:self.view.leftAnchor
         multiplier:2],
        [self.view.rightAnchor
         constraintEqualToSystemSpacingAfterAnchor:stackView.rightAnchor
         multiplier:2],
        [stackView.topAnchor
         constraintEqualToSystemSpacingBelowAnchor:self.view.topAnchor
         multiplier:2],
    ]];
    
    self.activityIndicator = [[SPLoadingView alloc] initWithFrame:CGRectMake(0, 0, 35, 35)];;
    self.activityIndicator.center = self.view.center;
    self.activityIndicator.lineColor = [UIColor systemBlueColor];
    [self.view addSubview:self.activityIndicator];
}

- (void)pay {
    
   self.paymentAmount = [[self.amountTextField.text substringFromIndex:1] stringByReplacingOccurrencesOfString:@"," withString:@""];

  [self.activityIndicator startAnimation];

  [[SPAPIClient getSharedInstance] createPaymentMethodWithType:@"CREDIT_CARD"
      account:self.cardTextField.cardNumber
      expDate:self.cardTextField.formattedExpirationDate
      cvv:self.cardTextField.cvc
      accountType:nil
      routing:nil
      pin:nil
      address:self.billingAddress
      address2:self.billingAddress2
      city:self.billingCity
      country:self.billingCountry
      state:self.billingState
      zip:self.cardTextField.postalCode?:self.billingZip
      company:self.company
      email:self.email
      phone:self.phoneNumber
      name:self.name
      nickname:self.nickname
      verification : self.isVerification
      success:^(SPPaymentMethod *paymentMethod) {
      
        if ([self.delegate respondsToSelector:@selector(paymentViewController:paymentMethodSuccess:)]) {
           [self.delegate paymentViewController:self paymentMethodSuccess:paymentMethod];
        }
      
        [[SPAPIClient getSharedInstance]
            createChargeWithToken:paymentMethod.token
            cvv:self.cardTextField.cvc
            capture: TRUE
            currency:nil
            amount:self.paymentAmount
            taxAmount:nil
            taxExempt: FALSE
            tip:nil
            surchargeFeeAmount:nil
            scheduleIndicator:nil
            description:self.paymentDescription
            order:nil
            orderId:self.orderId
            poNumber:nil
            metadata:self.paymentMetadata
            descriptor:nil
            txnEnv:nil
            achType:nil
            credentialIndicator:nil
            transactionInitiation:nil
            idempotencyKey: self.idempotencyKey
            needSendReceipt:self.isNeedSendReceipt
            success:^(SPCharge *charge) {
              [self.activityIndicator stopAnimation];
              if ([self.delegate respondsToSelector:@selector(paymentViewController:chargeSuccess:)]) {
                 [self.delegate paymentViewController:self chargeSuccess:charge];
              }
            }
            failure:^(SPError *error) {
              [self.activityIndicator stopAnimation];
              if ([self.delegate respondsToSelector:@selector(paymentViewController:chargeError:)]) {
                 [self.delegate paymentViewController:self chargeError:error];
              }
            }];
      }
      failure:^(SPError *error) {
      [self.activityIndicator stopAnimation];
      
      if ([self.delegate respondsToSelector:@selector(paymentViewController:paymentMethodError:)]) {
          [self.delegate paymentViewController:self paymentMethodError:error];
      }
  }];
}


-(BOOL)textField:(UITextField *)textField shouldChangeCharactersInRange:(NSRange)range replacementString:(NSString *)string{
    
    NSString *cleanCentString = [[textField.text componentsSeparatedByCharactersInSet: [[NSCharacterSet decimalDigitCharacterSet] invertedSet]] componentsJoinedByString:@""];
    NSInteger centValue = [cleanCentString intValue];
    NSNumberFormatter *f = [[NSNumberFormatter alloc] init];
    NSNumber *myNumber = [f numberFromString:cleanCentString];
    NSNumber *result;
    
    if ([textField.text length] < 16) {
        if (string.length > 0) {
            centValue = centValue * 10 + [string intValue];
            double intermediate = [myNumber doubleValue] * 10 +  [[f numberFromString:string] doubleValue];
            result = [[NSNumber alloc] initWithDouble:intermediate];
        } else {
            centValue = centValue / 10;
            double intermediate = [myNumber doubleValue]/10;
            result = [[NSNumber alloc] initWithDouble:intermediate];
        }
        textField.text = [self formatedValue:result];

        return FALSE;
    } else {
        return FALSE;
    }
    return TRUE;
}

- (NSString *)formatedValue:(NSNumber*)myNumber {
    NSNumber *formatedValue;
    formatedValue = [[NSNumber alloc] initWithDouble:[myNumber doubleValue] / 100.0f];
    NSNumberFormatter *_currencyFormatter = [[NSNumberFormatter alloc] init];
    _currencyFormatter.locale = [NSLocale localeWithLocaleIdentifier:@"en_US"];
    [_currencyFormatter setNumberStyle:NSNumberFormatterCurrencyStyle];
    return  [_currencyFormatter stringFromNumber:formatedValue];
}

@end

File: SeamlessPayCore/Classes/SPPaymentViewController.swift
// **
// * Copyright (c) Seamless Payments, Inc.
// *
// * This source code is licensed under the MIT license found in the
// * LICENSE file in the root directory of this source tree.
// *

import UIKit
import Foundation

public protocol SPPaymentViewControllerDelegate: AnyObject {
  func paymentViewControllerChargeSuccess(
    _ viewController: SPPaymentViewController,
    charge: Charge
  )
  func paymentViewControllerChargeError(
    _ viewController: SPPaymentViewController,
    error: SeamlessPayError
  )
  func paymentViewControllerPaymentMethodSuccess(
    _ viewController: SPPaymentViewController,
    paymentMethod: PaymentMethod
  )
  func paymentViewControllerPaymentMethodError(
    _ viewController: SPPaymentViewController,
    error: SeamlessPayError
  )
}

public class SPPaymentViewController: UIViewController {
  public var paymentType: PaymentType = .creditCard
  public var paymentDescription: String?
  public var paymentAmount: Decimal = 0.0
  public var billingAddress: Address?
  public var logoImage: UIImage?
  public weak var delegate: SPPaymentViewControllerDelegate?

  var cardTextField = SPPaymentCardTextField()
  var payButton: UIButton?
  var amountLabel: UILabel?
  var descriptionLabel: UILabel?
  var activityIndicator = SPLoadingView(frame: CGRect(x: 0, y: 0, width: 35, height: 35))
  var isEditAmount: Bool = false
  var isVerification: Bool = false
  var idempotencyKey: String?
  var orderId: String?
  var name: String?
  var paymentMetadata: String?

  private var formatter: NumberFormatter = {
    let formatter = NumberFormatter()
    formatter.numberStyle = .decimal
    formatter.minimumFractionDigits = 2
    formatter.maximumFractionDigits = 2
    formatter.locale = Locale(identifier: "en_US")

    return formatter
  }()

  override public func viewDidLoad() {
    super.viewDidLoad()

    view.backgroundColor = UIColor.white

    cardTextField.postalCodeEntryEnabled = true
    cardTextField.countryCode = "US"

    let button = UIButton(type: .custom)
    button.layer.cornerRadius = 5
    button.backgroundColor = UIColor.systemBlue
    button.titleLabel?.font = UIFont.systemFont(ofSize: 22)
    button.setTitle("Pay", for: .normal)
    button.addTarget(self, action: #selector(pay), for: .touchUpInside)
    payButton = button

    let amountLabel = UILabel(frame: CGRect(x: 0, y: 0, width: 250, height: 30))
    amountLabel.isUserInteractionEnabled = isEditAmount
    amountLabel.textColor = UIColor.darkGray
    let formattedAmount = formattedCurrency(paymentAmount)
    amountLabel.text = "Amount: \(formattedAmount)"
    self.amountLabel = amountLabel

    let descriptionLabel = UILabel()
    descriptionLabel.text = paymentDescription
    descriptionLabel.font = UIFont.systemFont(ofSize: 13)
    descriptionLabel.textColor = UIColor.gray
    descriptionLabel.numberOfLines = 2
    descriptionLabel.sizeToFit()
    self.descriptionLabel = descriptionLabel

    let logoImageView = UIImageView(image: logoImage)
    logoImageView.contentMode = .center

    let stackView = UIStackView(arrangedSubviews: [
      logoImageView,
      amountLabel,
      descriptionLabel,
      cardTextField,
      button,
    ])
    stackView.axis = .vertical
    stackView.translatesAutoresizingMaskIntoConstraints = false
    stackView.spacing = 20
    view.addSubview(stackView)

    NSLayoutConstraint.activate([
      stackView.leftAnchor.constraint(
        equalToSystemSpacingAfter: view.leftAnchor,
        multiplier: 2
      ),
      view.rightAnchor.constraint(
        equalToSystemSpacingAfter: stackView.rightAnchor,
        multiplier: 2
      ),
      stackView.topAnchor.constraint(
        equalToSystemSpacingBelow: view.topAnchor,
        multiplier: 2
      ),
    ])

    if let activityIndicator {
      activityIndicator.center = view.center
      activityIndicator.lineColor = UIColor.systemBlue
      view.addSubview(activityIndicator)
    }
  }

  private func formattedCurrency(_ value: Decimal) -> String {
    let formattedValue = NSDecimalNumber(decimal: value)
    formatter.numberStyle = .currency
    return formatter.string(from: formattedValue) ?? ""
  }

  private func formattedParameter(_ value: Decimal) -> String {
    let formattedValue = NSDecimalNumber(decimal: value)
    formatter.numberStyle = .decimal
    return formatter.string(from: formattedValue) ?? ""
  }
}

private extension SPPaymentViewController {
  @objc func pay() {
    activityIndicator?.startAnimation()

    APIClient.shared.tokenize(
      paymentType: paymentType,
      accountNumber: cardTextField.cardNumber ?? "",
      expDate: .init(month: cardTextField.expirationMonth, year: cardTextField.expirationYear),
      cvv: cardTextField.cvc,
      accountType: nil,
      routing: nil,
      pin: nil,
      billingAddress: billingAddress,
      name: name
    ) { result in

      switch result {
      case let .success(paymentMethod):
        DispatchQueue.main.async {
          self.delegate?.paymentViewControllerPaymentMethodSuccess(
            self,
            paymentMethod: paymentMethod
          )
        }

        APIClient.shared.createCharge(
          token: paymentMethod.token,
          cvv: self.cardTextField.cvc,
          capture: true,
          currency: nil,
          amount: self.formattedParameter(self.paymentAmount),
          taxAmount: nil,
          taxExempt: false,
          tip: nil,
          surchargeFeeAmount: nil,
          description: self.paymentDescription,
          order: nil,
          orderID: self.orderId,
          poNumber: nil,
          metadata: self.paymentMetadata,
          descriptor: nil,
          entryType: nil,
          idempotencyKey: self.idempotencyKey
        ) { result in

          DispatchQueue.main.async {
            self.activityIndicator?.stopAnimation()

            switch result {
            case let .success(charge):
              self.delegate?.paymentViewControllerChargeSuccess(self, charge: charge)
            case let .failure(error):
              self.delegate?.paymentViewControllerChargeError(self, error: error)
            }
          }
        }
      case let .failure(error):
        DispatchQueue.main.async {
          self.activityIndicator?.stopAnimation()
          self.delegate?.paymentViewControllerPaymentMethodError(self, error: error)
        }
      }
    }
  }
}
